### ISSUE TITLE
Bug Sighting - Looks like parse error

### ISSUE THREAD

**Issue by RobertESinger** [sentry07/netlinx-libJSON#1]
Hi -

Accessing api.openweathermap.org from Netlinx program

The program does a recursive traversal starting with the top level json object returned by JSON_ParseObject ( ) of the data returned from

api.openweathermap.org/data/2.5/onecall?lat=33.45&lon=-112.07&exclude=hourly,minutely,alerts&appid=275191df4cf9dd0e08f3c410021a6033

At some point in the processing, the parse fails with 
[JSON] Validator returned the following error: ERROR at position 72: End of array unexpected.

The object being parsed is
{"id":804,"main":"Clouds","description":"overcast clouds","icon":"04d"}],"clouds":100,"pop":0,"uvi":2.39}

Clearly, the leading '[' is missing.

However, the data stream does actually contain the '['
[{"id":804,"main":"Clouds","description":"overcast clouds","icon":"04d"}]

So, something is already wrong with the parse before getting to this particular object.

This error seems to occur frequently with data returned by various calls to the api.openweathermap.org website.

How would you like to proceed on further diagnosis?

Bob

**Comment by RobertESinger** [comment]
I also noticed that some object values have a '}' at the end.  I don't think this is correct.  Examples:
coord.lat 33.45}
...
main.humidity 10}
...
wind.deg 174}

**Comment by sentry07** [comment]
Can you share a snippet of code where you're making these calls?

**Comment by RobertESinger** [comment]
cmd = "'GET /data/2.5/weather?q=Phoenix,Arizona&appid=275191df4cf9dd0e08f3c410021a6033', ' HTTP/1.1', ascii_CR, ascii_LF,
				 'Host: api.openweathermap.org', ascii_CR, ascii_LF,
				 'Content-Type: application/json', ascii_CR, ascii_LF,
				 'Connection: close', ascii_CR, ascii_LF,
				 ascii_CR, ascii_LF"

	traceMsg ( "routineName, 'sending cmd: ', cmd" )

	SEND_STRING dev_Weather, "cmd"
...

		REMOVE_STRING ( weatherDataBuf, "ascii_CR, ascii_LF, ascii_CR, ascii_LF", 1 )

		JSON_ParseObject ( weatherDataBuf, jsonWeatherObject )
		traverseObject ( jsonWeatherObject, '' )

...



DEFINE_FUNCTION traverseObject ( _JSON_Object jsonObj, CHAR name [ ] ) {

	STACK_VAR CHAR routineName [ 32 ]
	STACK_VAR INTEGER i
	STACK_VAR CHAR newName [ 64 ]
	STACK_VAR _JSON_Array a
	STACK_VAR _JSON_Object o


	routineName = 'traverseObject ( ):  '


	traceMsg ( "routineName" )

	FOR ( i = 1; i <= jsonObj.Count; i ++ ) {

		newName = "name, jsonObj.KV [ i ].K"

		SWITCH ( jsonObj.KV [ i ].nType ) {

			CASE _JSON_IsObject: {
				traceMsg ( "routineName, newName" )
				JSON_ParseObject ( jsonObj.KV [ i ].V, o )
				traverseObject ( o, "newName, '.'" )
			}

			CASE _JSON_IsArray: {
				traceMsg ( "routineName, newName" )
				JSON_ParseArray ( jsonObj.KV [ i ].V, a )
				traverseArray ( a, "newName, '.'" )
			}

			CASE _JSON_IsString: {
				traceMsg ( "routineName, newName, ' "', jsonObj.KV [ i ].V, '"'" )
				insertionSort ( newName, jsonObj.KV [ i ].V, jsonObj.KV [ i ].nType )
//				listJSONData ( )
			}

			CASE _JSON_IsNumber:
			CASE _JSON_IsBool:
			CASE _JSON_IsNull: {
				traceMsg ( "routineName, newName, ' ', jsonObj.KV [ i ].V" )
				insertionSort ( newName, jsonObj.KV [ i ].V, jsonObj.KV [ i ].nType )
//				listJSONData ( )
			}

		}

	}

		traceMsg ( "routineName, 'return'" )

}


DEFINE_FUNCTION traverseArray ( _JSON_Array jsonAry, CHAR name [ ] ) {

	STACK_VAR CHAR routineName [ 32 ]
	STACK_VAR INTEGER i
	STACK_VAR CHAR newName [ 64 ]
	STACK_VAR _JSON_Array a
	STACK_VAR _JSON_Object o


	routineName = 'traverseArray ( ):  '

	traceMsg ( "routineName" )

	FOR ( i = 1; i <= jsonAry.Count; i ++ ) {

		newName = "name, ITOA ( i )"

		SWITCH ( jsonAry.List [ i ].nType ) {

			CASE _JSON_IsObject: {
				traceMsg ( "routineName, newName" )
				JSON_ParseObject ( jsonAry.List [ i ].Value, o )
				traverseObject ( o, "newName, '.'" )
			}

			CASE _JSON_IsArray: {
				traceMsg ( "routineName, newName" )
				JSON_ParseArray ( jsonAry.List [ i ].Value, a )
				traverseArray ( a, "newName, '.'" )
			}

			CASE _JSON_IsString: {
				traceMsg ( "routineName, newName, ' "', jsonAry.List [ i ].Value, '"'" )
				insertionSort ( newName, jsonAry.List [ i ].Value, jsonAry.List [ i ].nType )
//				listJSONData ( )
			}

			CASE _JSON_IsNumber:
			CASE _JSON_IsBool:
			CASE _JSON_IsNull: {
				traceMsg ( "routineName, newName, ' ', jsonAry.List [ i ].Value" )
				insertionSort ( newName, jsonAry.List [ i ].Value, jsonAry.List [ i ].nType )
//				listJSONData ( )
			}

		}

	}

	traceMsg ( "routineName, 'return'" )

}

**Comment by RobertESinger** [comment]
Hi -

Other than cutting off the HTML header, no manipulation of the returned json data is done.

As the data is recursively parsed in traverseObject ( ) and traverseArray ( ), the keys, values and types are stored in another structure for later use.

Let me know if you need anything else.

Bob

**Comment by sentry07** [comment]
I mocked up your code (other than the insertion sort) and ran it on my NI-3100 and found the bug with returning the closing } on the integer value. I'm not seeing a bug with the validation not working. I'm getting a full parse with no errors. I would recommend in the traverse functions to use JSON_ParseValidObject since that will bypass the validation function and save some CPU usage. 

My event setup:
![image](https://user-images.githubusercontent.com/6489018/103165304-cfc9b200-47ca-11eb-9984-54a53926f414.png)

My data parsing output:
![image](https://user-images.githubusercontent.com/6489018/103165275-7eb9be00-47ca-11eb-8b9d-057443bc1bbd.png)

**Comment by sentry07** [comment]
I committed an update with that number bugfix but will leave this issue open until we figure out why the other problem is happening.

**Comment by RobertESinger** [comment]
OK.  Your fix for not returning a trailing '}' works for me.  Thanks for that.

Before we delve into the other issue, I'd like to go on a digression about the json parsing routines data structure constant values.

So, it starts with an HTTP call that returns lots of data (~8-16k).  My program swallows this into an appropriately-sized buffer.

For example, try the following URL:

GET /data/2.5/onecall?lat=33.45&lon=-112.07&exclude=minutely,alerts&units=imperial&mode=json&lang=en&appid=275191df4cf9dd0e08f3c410021a6033 HTTP/1.1$0D$0AHost: api.openweathermap.org$0D$0AContent-Type: application/json$0D$0AConnection: close$0D$0A$0D$0A

True or false:  The program conatant _JSON_InputSize must be the same size as the program's buffer that holds the returned data.

Can you give guidance for the sizes of the constants _JSON_MaxPairs, _JSON_KeySize, _JSON_ValueSize & _JSON_ArraySize?  Or is it just trial and error to determine the values?

I suppose I could make them all very large, and then, after a successful parse, look at the sizes of the resultant data, and set them accordingly.

Bob

**Comment by sentry07** [comment]
_JSON_InputSize must be larger than the expected full JSON string.
_JSON_MaxPairs must be larger than the expected number of key-value pairs of any node in the tree.
_JSON_KeySize must be larger than the expected length of any key name in the tree.
_JSON_ValueSize must be larger than the expected length of any value in the tree. This one is tricky because you can have a full tree as a value so it's expected to be a large size.
_JSON_ArraySize must be larger than the expected length of any array in the tree.

All of these will vary based on what you're expecting to parse. If it's a known source with known length like your weather data, then you can bring down the numbers of any of those to cut down on memory usage.

**Comment by RobertESinger** [comment]
More info on the initial problem report:

Once the top level object has been parsed successfully with JSON_ParseObject, you recommended using JSON_ParseValidObject instead to save compute time.  The reason why my program trips over some errors is that parsing the next object or array is based on the previously extracted Value.  If the previously extracted Value isn't syntactically correct, JSON_ParseValidObject won't squawk.  By using JSON_ParseObject, it uncovers that a previously extracted Value isn't correct.

For example:

 object {"pod":"d"},"dt_txt":"2020-12-30 00:00:00"}
 [JSON] Validator returned the following error: ERROR at position 43: End of object unexpected.

This is why, in my initial problem report, I found '}' characters at the end of some objects.  The previously extracted Value was incorrect.  The specific parsing bugs might not be the same, but the result is a bad extracted Value.

I'm currently testing with the URL:

http://api.openweathermap.org/data/2.5/forecast?q=Phoenix,Arizona&cnt=5&units=imperial&mode=json&lang=en&appid=275191df4cf9dd0e08f3c410021a6033

I will resume using JSON_ParseObject, as it appears to find variations on the same problem.

Let me know if you need more info.

Bob

**Comment by RobertESinger** [comment]
On the array size constants issue, I embedded some element count and size accumulators in my program to help determine which arrays are being overrun.  At the end of a run, it prints the accumulated values so I will know which constant values to (not) fiddle with. 

This would be much easier to gracefully handle if Netlinx had a language feature like dynamic memory allocation.

**Comment by RobertESinger** [comment]
OK - To distill the problem space:  Under certain circumstances, the Value isn't being correctly extracted and assigned to the .V field (for objects) and .Value fields (for arrays).

Using this URL:

http://api.openweathermap.org/data/2.5/forecast?q=Phoenix,Arizona&units=imperial&mode=json&lang=en&appid=275191df4cf9dd0e08f3c410021a6033

I added SEND_STRINGs to your code in JSON_ParseArray and JSON_ParseObject and, in the specific case of the URL above, found the assigned value is incorrect in this section of code in JSON_ParseArray:

			CASE '}':						// End of object
			{
				IF (nJSONDebug) SEND_STRING 0,"'[JSON] I see a } at ',ITOA(F1),' (bInArray: ',ITOA(bInArray),', bInChildObj: ',ITOA(bInChildObj),')'"
				IF (!bInArray)
				{
					BrChildObj --
					IF (BrChildObj == 0)			// This is the end of the child object
					{
						nFind = F1
						jTempArray.List[nValueCount].nType = _JSON_IsObject
						jTempArray.List[nValueCount].Value = MID_STRING(cTempInput,nChildObjStart,nFind-(nChildObjStart-1))
					send_string 0, "'A.2 --> value = ', jTempArray.List[nValueCount].Value"
						nChildObjStart = 0
						bInChildObj = 0
					}
				}
			}

The parameters for MID_STRING aren't correct.

In the first problem report comment I sent, using a different URL yielded the same problem, but regarding the unexpected end of an array. 

Both problem instances point to specific areas of your code for you to examine further.

Thx - Bob

**Comment by sentry07** [comment]
Think I got the problem fixed. Please check the latest version of the file.

**Comment by RobertESinger** [comment]
I brought in the latest version of the file and get the following new error:

[JSON] Array not enclosed in brackets.

**Comment by sentry07** [comment]
Check your size constants. In order to get that last URL you posted to work, I had to set the _JSON_Value size to 18000 and the _JSON_InputSize to 20000.

**Comment by RobertESinger** [comment]
LOL - My mistake.  I wasn't quite awake yet when I tested, and forgot to bring across the updated structure sizes.

After getting appropriately caffeinated, I ran 6 tests against various data from the same website, and everything appeared to work.

Thanks for that.

At this time, I don't see any additional work required for this issue. 

I'm going to continue testing for the next few days.  If everything continues as it did today, I'll close this issue.

Bob
